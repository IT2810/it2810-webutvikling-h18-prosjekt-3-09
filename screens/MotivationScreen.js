import React from 'react';
import { 
    Text,
    View,
    ListView,
    StyleSheet,
    Modal,
    Button,
    TouchableHighlight,
} from 'react-native';
import { TextInput } from 'react-native-gesture-handler';
import { withStore } from "../components/Store/"
import {sendNotification} from "../components/Notification"

const tempData = [];  // Load with asyncstorage

class MotivationScreen extends React.Component {
  static navigationOptions = {
    title: 'Motivation Manager',
};



  constructor() {
    super();
    this.ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2});
    this.state = {
      managerData: [],
      dataSource: this.ds.cloneWithRows([]),     // TODO: initialize empty
      modalVisible: false,   
      modalInput: '',
      editGoal: false,
      editId: -1,
    }; 
  }


  async componentDidMount() {
      try {
          const savedData = await this.props.actions.getItem("managerData") // datakey is either string of 'todo' or 'manager', essentially the key of the data where you store it in AsyncStorage
          this.setState({managerData: savedData || []}) // If no data was found in AsyncStorage, set an empty array
          tempData = savedData || [];
          console.log("successfully mounted")
    } catch (error) {
      sendNotification("error", error) // To give error feedback
    }
 }

 // Use this function to update both the local state, and the database at the same time.
 updateData = async data => {
    try {
      this.setState({data: data})
      await this.props.actions.setItem("managerData", data)
      this.updateListRows();
    } catch (error) {
      sendNotification("error", error)
    }
 }


  deleteElement = () => {
    tempData.splice(this.state.editId, 1);
    this.setState({
        modalInput: '',
        editGoal: false,
        editId: -1,
    });
    this.updateData(tempData);
    this.setModalInvisible();
  }

  onPressNewGoal = () => {
    this.setState({
        modalVisible: true
    });
  }

  cancelModal = () => {
    if(tempData != [])
        this.setState({
            modalInput: '',
            editGoal: false,
            editId: -1,
            dataSource: this.ds.cloneWithRows(tempData)
        });
    this.setModalInvisible();
  }


  updateListRows = () => {
      this.setState({
        dataSource: this.ds.cloneWithRows(tempData)
      });
  }

  setModalInvisible = () => {
    this.setState({modalVisible: false});
  }

    closeAndCreate = () => {
        if(this.state.editGoal) {
            tempData[this.state.editId] = this.state.modalInput;
        }
        else {
            tempData.push(this.state.modalInput);  // unshift()  prepend, but listview is bugged
        }
        
        this.updateData(tempData);
        
        this.setState({
            modalInput: '',
            editGoal: false,
            editId: -1,
        });
        this.setModalInvisible();
    }

    onRowPress = (rowID) => {
        this.setState({
            editGoal: true,
            editId: rowID,
            modalInput: tempData[rowID],
            modalVisible: true,
        });        
    }

    renderListRow = (rowData, sectionID, rowID) => {
        return (
            <TouchableHighlight
            onPress = {() => this.onRowPress(rowID)}
            style={styles.listElement}
            underlayColor="#33aaddaa"
            >
                <Text style={styles.listText}>{rowData}</Text>
            </TouchableHighlight>
        );
    }


  render() {
    return (
    <View style={styles.pageContainer}>
        <View style={styles.container}>

            <TouchableHighlight
                onPress={this.onPressNewGoal}
                style = {styles.newElementButton}

            >
            <Text style={styles.buttonText}>Legg til nytt m책l</Text>
            </TouchableHighlight>
        
            <ListView
            enableEmptySections={true}
            dataSource={this.state.dataSource}
            renderRow={this.renderListRow}
            style={styles.listElementContainer}
            />
        </View>
        <Modal
            transparent={true}
            animationType = 'slide' // fade or none
            visible = {this.state.modalVisible}
            style = {styles.modal}
            onRequestClose={() => {
                console.log("Request to close modal");
                this.cancelModal
            }}>
            <View style={styles.modal}>
                <Text style={styles.modalTitle}>{this.state.editGoal ? 'Endre m책l' : 'Nytt m책l'}</Text>
                <TextInput
                    style={styles.modalInput}
                    placeholder = 'M책lbeskrivelse'
                    onChangeText={(modalInput) => this.setState({modalInput})}
                    value={this.state.modalInput}/>
                
                <View style={styles.modalButtons}>
                    <TouchableHighlight
                        onPress={this.closeAndCreate}
                        style = {styles.modalCreateButton}
                        underlayColor="#00000022">
                        <Text style={styles.buttonText}>{this.state.editGoal ? 'Lagre' : 'Opprett'}</Text>
                    </TouchableHighlight>
                    <TouchableHighlight
                        onPress={this.cancelModal}
                        style = {styles.modalCancelButton}
                        underlayColor="#aaaaaaaa">
                        <Text style={styles.buttonText}>Avbryt</Text>
                    </TouchableHighlight>
                {this.state.editGoal && <Button
                onPress={this.deleteElement}
                style={styles.modalDeleteButton}
                title="Slett"
                />}
                </View>

            </View>
        </Modal>


    </View>
    );
  }
}

export default withStore(MotivationScreen)

const styles = StyleSheet.create({
    pageContainer: {
        flex: 1,
        flexDirection: 'column',
        justifyContent: 'space-between',
        backgroundColor: '#fff',
    },
    container: {
      backgroundColor: '#fff',
      marginHorizontal: 20,
    }, 
    newElementButton: {
       backgroundColor: "rgba(40,200,40, 1)",
       paddingVertical: 20,
       marginVertical: 10,
       borderRadius: 10,
       elevation: 3,    // android only
    }, 
    listElementContainer: {
        marginBottom: 100,
    }, 

    listElement: {
        flex: 1,
        borderWidth: 2,
        flexDirection: 'row',
        backgroundColor: '#EEEEEE',
        borderStyle: 'solid',
        borderColor: '#DDDDDD',
        borderRadius: 5,
        height: 50,
        marginBottom: 5,
        
        
    }, 
    listText: {
        fontSize: 20,
        textAlignVertical: 'center',
        paddingLeft: 30,
    }, 
    modal: {
        flex: 1,
        backgroundColor: "#ffffff",
        marginHorizontal: 20,
        marginVertical: 25,
        elevation: 20, // android only
    },
    modalButtons: {
        marginHorizontal: 20,
        flexDirection: 'row',
        justifyContent: 'space-between',
        height: 100,
    },
    modalInput: {
        marginHorizontal: 30,
        borderRadius: 5,
        height: 30,
        paddingHorizontal: 5,
        marginBottom: 30,
        marginVertical: 20,
        paddingBottom: 5,
        fontSize: 20,
    },
    modalTitle: {
        alignItems: 'center',
        textAlign: 'center',
        fontSize: 16,
    },
    buttonText: {
        fontSize: 24,
        color: "#fff",
        textAlign: 'center',
    },
    modalCreateButton: {
        backgroundColor: "rgba(40,200,40, 1)",
        paddingVertical: 20,
        borderRadius: 10,
        width: 120,
        elevation: 3,    // android only
    },
    modalCancelButton: {
        backgroundColor: "rgba(255,200,40, 1)",
        paddingVertical: 20,
        borderRadius: 10,
        width: 120,
        elevation: 3,    // android only
    },
    modalDeleteButton: {
        backgroundColor: "rgba(200, 40,40, 1)",
        paddingVertical: 20,
        borderRadius: 10,
        elevation: 3,    // android only
    },
        
});