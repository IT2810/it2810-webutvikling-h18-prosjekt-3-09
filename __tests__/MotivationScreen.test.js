import React from "react";
import MotivationScreen from "../screens/MotivationScreen";
import renderer from "react-test-renderer";

jest.mock("react-native-gesture-handler", () => {});

describe("Testing MotivationScreen", () => {
  const tree = renderer.create(<MotivationScreen />);

  test("renders correctly", () => {
    expect(tree.toJSON()).toMatchSnapshot();
  });

  describe("Initial state", () => {
    test("modal should be hidden", () => {
      expect(tree.toTree().rendered.instance.state.modalVisible).toBe(false);
    });

    test("edit view should be closed", () => {
      expect(tree.toTree().rendered.instance.state.editGoal).toBe(false);
    });

    test("input should be empty", () => {
      expect(tree.toTree().rendered.instance.state.modalInput).toEqual("");
    });

    test("no goal should be selected for edit", () => {
      expect(tree.toTree().rendered.instance.state.editId).toEqual(-1);
    });
  });

  describe("Modal functions", () => {
    describe("Reset selected modal", () => {
      test("clear input successfully", () => {
        tree.toTree().rendered.instance.resetModalState();
        expect(tree.toTree().rendered.instance.state.modalInput).toEqual("");
      });

      test("edit mode exited", () => {
        tree.toTree().rendered.instance.resetModalState();
        expect(tree.toTree().rendered.instance.state.editGoal).toBe(false);
      });

      test("deselect goal", () => {
        tree.toTree().rendered.instance.resetModalState();
        expect(tree.toTree().rendered.instance.state.editId).toEqual(-1);
      });
    });

    test("modal opens successfully", () => {
      tree.toTree().rendered.instance.onPressNewGoal();
      expect(tree.toTree().rendered.instance.state.modalVisible).toBe(true);
    });

    test("modal closes successfully", () => {
      tree.toTree().rendered.instance.setModalInvisible();
      expect(tree.toTree().rendered.instance.state.modalVisible).toBe(false);
    });
  });

  describe("Handle goals", () => {
    test("create goal successfully", () => {
      tree.toTree().rendered.instance.handleInputChange("Walk");
      tree.toTree().rendered.instance.closeAndCreate();
      expect(tree.toTree().rendered.instance.state.data).toEqual(["Walk"]);
    });

    describe("Select a goal", () => {
      test("enter edit mode", () => {
        tree.toTree().rendered.instance.onRowPress(0);
        expect(tree.toTree().rendered.instance.state.editGoal).toBe(true);
      });

      test("select correct goal", () => {
        expect(tree.toTree().rendered.instance.state.editId).toEqual(0);
      });

      test("input should be current goal", () => {
        expect(tree.toTree().rendered.instance.state.modalInput).toEqual(
          "Walk"
        );
      });
    });

    test("cancel modal succesfully", () => {
      tree.toTree().rendered.instance.cancelModal();
      expect(tree.toTree().rendered.instance.state.modalVisible).toBe(false);
    });

    test("delete goal successfully", () => {
      tree.toTree().rendered.instance.deleteElement();
      expect(tree.toTree().rendered.instance.state.data).toHaveLength(0);
    });
  });
});
