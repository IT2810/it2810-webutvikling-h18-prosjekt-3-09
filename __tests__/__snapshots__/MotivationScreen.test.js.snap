// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing MotivationScreen renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
      "flexDirection": "column",
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "#fff",
        "marginHorizontal": 20,
      }
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "rgba(40,200,40, 1)",
          "borderRadius": 10,
          "elevation": 3,
          "marginVertical": 10,
          "paddingVertical": 20,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#fff",
            "fontSize": 24,
            "textAlign": "center",
          }
        }
      >
        Legg til nytt mål
      </Text>
    </View>
  </View>
  <Modal
    animationType="slide"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    style={
      Object {
        "backgroundColor": "#ffffff",
        "elevation": 20,
        "flex": 1,
        "marginHorizontal": 20,
        "marginVertical": 25,
      }
    }
    transparent={true}
    visible={false}
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "elevation": 20,
          "flex": 1,
          "marginHorizontal": 20,
          "marginVertical": 25,
        }
      }
    >
      <Text
        accessible={true}
        allowFontScaling={true}
        ellipsizeMode="tail"
        style={
          Object {
            "alignItems": "center",
            "fontSize": 16,
            "textAlign": "center",
          }
        }
      >
        Nytt mål
      </Text>
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Målbeskrivelse"
        style={
          Object {
            "borderRadius": 5,
            "fontSize": 20,
            "height": 30,
            "marginBottom": 30,
            "marginHorizontal": 30,
            "marginVertical": 20,
            "paddingBottom": 5,
            "paddingHorizontal": 5,
          }
        }
        value=""
      />
      <View
        style={
          Object {
            "flexDirection": "row",
            "height": 100,
            "justifyContent": "space-between",
            "marginHorizontal": 20,
          }
        }
      >
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(40,200,40, 1)",
              "borderRadius": 10,
              "elevation": 3,
              "paddingVertical": 20,
              "width": 120,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#fff",
                "fontSize": 24,
                "textAlign": "center",
              }
            }
          >
            Opprett
          </Text>
        </View>
        <View
          accessibilityComponentType={undefined}
          accessibilityLabel={undefined}
          accessibilityTraits={undefined}
          accessible={true}
          hasTVPreferredFocus={undefined}
          hitSlop={undefined}
          isTVSelectable={true}
          nativeID={undefined}
          onLayout={undefined}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "backgroundColor": "rgba(255,200,40, 1)",
              "borderRadius": 10,
              "elevation": 3,
              "paddingVertical": 20,
              "width": 120,
            }
          }
          testID={undefined}
          tvParallaxProperties={undefined}
        >
          <Text
            accessible={true}
            allowFontScaling={true}
            ellipsizeMode="tail"
            style={
              Object {
                "color": "#fff",
                "fontSize": 24,
                "textAlign": "center",
              }
            }
          >
            Avbryt
          </Text>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
